playgame.py

```diff
@@ -16,40 +18,52 @@
+    args.players = validate_players(args.players)
     if args.is_tournament:
         run_tournament(args)
     else:
         run_one_game(args)
 
+def validate_players(players):
+    return [player for player in players if locate(player) is not None and
+            isinstance(locate(player)(), HanabiPlayer)]
+
 def run_tournament(args):
-    tournament_scores = dict.fromkeys(args.players, 0)
+    tournament_scores = dict.fromkeys(args.players, [])
     pairings = list(itertools.combinations(args.players, 2))
+    disqualified = []
     for player1, player2 in pairings:
-        try:
-            game = HanabiGame([player1, player2], args.seed, HanabiVariant(args.variant))
-            game.play_game(args)
-            score = game.table.score()
-        except:
-            #if a player has messed up, penalize both
-            #could diqualify the failing bot if we can determine who failed
-            logger.warning('{player1} or {player2} failed to complete a game'.format(
-                player1 = player1,
-                player2 = player2
-                ))
-            score = -25
-        finally:
-            for handler in logger.handlers:
-                if handler.__class__ is RotatingFileHandler:
-                    handler.doRollover()
-            tournament_scores[player1] += score
-            tournament_scores[player2] += score
-    tournament_scores = {k: v / len(pairings) for k, v in tournament_scores.iteritems()}
+        if player1 and player2 not in disqualified:
+            try:
+                game = HanabiGame([player1, player2], args.seed, HanabiVariant(args.variant))
+                game.play_game(args)
+                score = game.table.score()
+                for handler in logger.handlers:
+                    if handler.__class__ is RotatingFileHandler:
+                        handler.doRollover()
+                tournament_scores[player1].append(score)
+                tournament_scores[player2].append(score)
+            except InvalidHanabiMoveException as err:
+                logger.error(err.message)
+                disqualified.append(player1 if err.player_id == 0 else player2)
+                tournament_scores = {k : v for k, v in tournament_scores.iteritems() if k not in disqualified}
+                logger.warning('removed player {player} from tournament'.format(player = disqualified))
+    tournament_results = {key: [numpy.mean(val), numpy.var(val)] for key, val in tournament_scores.iteritems()}
     logger.info('Scores: {scores}'.format(scores = tournament_scores))
-    winning_score = max(tournament_scores.itervalues())
-    winners = [key for key, value in tournament_scores.items() if value == winning_score]
+    logger.info('Results: {results}'.format(results = tournament_results))
+    winning_average = max(tournament_results.itervalues())
+    average_winners = {key: val for key, val in tournament_results.items() if val == winning_average}
```

Comment:
What do you think about moving winner calculations into a separate function?

It seems like there's enough going on with it that some tests might be nice as well, though it is essentially private to this class, which generally means it doesn't get tested.

Reply:
Yea, this function is becoming a monster. This is a very good suggestion